test/functional/apm.test.js

```diff
@@ -701,6 +701,45 @@ describe('APM', function() {
     }
   });
 
+  // NODE-3358
+  it('should not allow mutation of internal state from commands returned by event monitoring', function() {
+    const started = [];
+    const succeeded = [];
+    const client = this.configuration.newClient(
+      { writeConcern: { w: 1 } },
+      { maxPoolSize: 1, monitorCommands: true }
+    );
+    client.on('commandStarted', filterForCommands('insert', started));
+    client.on('commandSucceeded', filterForCommands('insert', succeeded));
+    let documentToInsert = { a: { b: 1 } };
+    return client
+      .connect()
+      .then(client => {
+        const db = client.db(this.configuration.db);
+        return db.collection('apm_test').insertOne(documentToInsert);
+      })
+      .then(r => {
+        expect(r)
+          .to.have.property('insertedId')
+          .that.is.an('object');
+        expect(started).to.have.lengthOf(1);
+        // Check if contents of returned document are equal to document inserted (by value)
+        expect(documentToInsert).to.deep.equal(started[0].command.documents[0]);
+        // Check if the returned document is a clone of the original. This confirms that the
+        // reference is not the same.
+        expect(documentToInsert !== started[0].command.documents[0]).to.equal(true);
+        expect(documentToInsert.a !== started[0].command.documents[0].a).to.equal(true);
+
+        started[0].command.documents[0].a.b = 2;
+        expect(documentToInsert.a.b).to.equal(1);
+
+        expect(started[0].commandName).to.equal('insert');
+        expect(started[0].command.insert).to.equal('apm_test');
+        expect(succeeded).to.have.lengthOf(1);
+        return client.close();
```

Comment:
sorry for not catching this in the 4.0 PR, but, since this belongs in the clean up, the ideal way to do it would be to have this in an afterEach block with the creation of the client in a beforeEach; to avoid refactoring too much, it's fine to wrap this in its own describe for that purpose